### PR TITLE
Using Method ABIs instead of ABI IDs

### PR DESCRIPTION
**Pull request by eberger727**

### CODE REVIEW

File: solidity/contracts/ERC20WithData.sol [hunk 1]
```diff
@@ -9,26 +9,25 @@ import '@openzeppelin/contracts/utils/Context.sol';
...
         _mint(to, amount);
     }
 
     function transferWithData(
-        address from,
         address to,
         uint256 amount,
         bytes calldata data
-    ) external virtual {
-        _transfer(from, to, amount);
```
Comment (awrichar) [inline solidity/contracts/ERC20WithData.sol:28]: Why did you change the signature here? I think we should keep the same parameters as the base method (`from, to, amount`) and pass them through.

Comment (awrichar) [reply solidity/contracts/ERC20WithData.sol:28]: Oh I see, you were calling a private method without approval checks.

Perhaps we should be proxying to `transferFrom` instead?

File: solidity/contracts/ERC20WithData.sol [hunk 2]
```diff
@@ -9,26 +9,25 @@ import '@openzeppelin/contracts/utils/Context.sol';
...
+    ) public {
+        _transfer(_msgSender(), to, amount);
     }
 
-    function burnWithData(address from, uint256 amount, bytes calldata data) external virtual {
-        _burn(from, amount);
```
Comment (awrichar) [inline solidity/contracts/ERC20WithData.sol:32]: Personally I would lean toward passing `from` here but requiring it to match `_msgSender()`. Let the logic be in the contract so that it's easier for users to replace the contract without needing to alter the connector to have their own burn policy.

File: src/tokens/tokens.interfaces.ts [hunk 3]
```diff
@@ -154,8 +195,9 @@ export class TokenTransfer {
   requestId?: string;
 }
 
-export class TokenMint extends OmitType(TokenTransfer, ['from']) {}
-export class TokenBurn extends OmitType(TokenTransfer, ['to']) {}
```
Comment (awrichar) [inline src/tokens/tokens.interfaces.ts:158]: I don't think these types should be changed. We should keep the connector interface the same - if certain mint/burn/transfer operations are not authorized, it should be decided by the smart contract.

File: src/tokens/tokens.controller.ts [hunk 4]
```diff
@@ -54,14 +52,14 @@ export class TokensController {
...
-  @HttpCode(204)
+  @HttpCode(200)
   @ApiOperation({
     summary: 'Activate a token pool to begin receiving transfer events',
     description: 'Will retrigger the token-pool event for this pool as a side-effect',
   })
   @ApiBody({ type: TokenPoolActivate })
   async activatePool(@Body() dto: TokenPoolActivate) {
-    await this.service.activatePool(dto);
+    return await this.service.activatePool(dto);
```
Comment (awrichar) [inline src/tokens/tokens.controller.ts:62]: You can remove `await` (just return the promise)

Comment (awrichar) [reply src/tokens/tokens.controller.ts:62]: And if you remove `await`, you can remove `async` on the function as well

File: src/tokens/tokens.controller.ts [hunk 5]
```diff
@@ -54,14 +52,14 @@ export class TokensController {
   }
 
   @Post('activatepool')
-  @HttpCode(204)
+  @HttpCode(200)
   @ApiOperation({
     summary: 'Activate a token pool to begin receiving transfer events',
     description: 'Will retrigger the token-pool event for this pool as a side-effect',
```
Comment (awrichar) [inline src/tokens/tokens.controller.ts:58]: Adjust these texts as well... no token-pool websocket event now, right?

File: src/tokens/tokens.service.ts [hunk 6]
```diff
@@ -129,101 +156,156 @@ export class TokensService {
...
-      this.http.post<EthConnectAsyncResponse>(
-        `${this.instanceUrl}/create`,
-        {
-          data: encodeHex(dto.data ?? ''),
-          name: dto.name,
-          symbol: dto.symbol,
-        },
-        this.postOptions(dto.operator, dto.requestId),
-      ),
+      this.http.get<EthConnectContractsResponse>(`${this.baseUrl}/contracts/${dto.config.address}`),
```
Comment (awrichar) [inline src/tokens/tokens.service.ts:166]: Should confirm that this works even if the contract was deployed outside of ethconnect.

File: src/tokens/tokens.interfaces.ts [hunk 7]
```diff
@@ -56,21 +95,23 @@ export class TokenPool {
...
   symbol: string;
 
+  @ApiProperty()
+  @IsNotEmpty()
+  type: TokenType;
+
   @ApiProperty({ description: contractConfigDescription })
   @IsOptional()
-  config?: any;
+  config: {
```
Comment (awrichar) [inline src/tokens/tokens.interfaces.ts:104]: If you annotate with `@IsOptional`, you should probably expect the item might be undefined. If it's required, you probably want to annotate with `@IsDefined` instead

File: src/tokens/tokens.interfaces.ts [hunk 8]
```diff
@@ -56,21 +95,23 @@ export class TokenPool {
...
+    address: string;
+  };
 
   @ApiProperty()
   @IsOptional()
-  data?: string;
+  data: any;
```
Comment (awrichar) [inline src/tokens/tokens.interfaces.ts:110]: I think FireFly guarantees this will be a string or undefined. Is there a reason to use `any`?

File: src/tokens/tokens.interfaces.ts [hunk 9]
```diff
@@ -176,10 +218,10 @@ class tokenEventBase {
   timestamp: string;
 
   @ApiProperty()
-  transaction: BlockchainTransaction;
+  transaction?: BlockchainTransaction;
 
   @ApiProperty()
-  type: TokenType;
+  type: string;
```
Comment (awrichar) [inline src/tokens/tokens.interfaces.ts:224]: Why change to string?

File: src/health/health.controller.ts [hunk 10]
```diff
@@ -24,13 +24,13 @@ export class HealthController {
...
         this.http.pingCheck(
           'ethconnect-instance',
-          this.tokensService.instanceUrl,
+          this.tokensService.baseUrl,
           basicAuth(this.tokensService.username, this.tokensService.password),
         ),
       () =>
         this.http.pingCheck(
           'ethconnect-contract',
-          this.tokensService.contractInstanceUrl,
```
Comment (awrichar) [inline src/health/health.controller.ts:33]: Any reason to have two duplicate ping checks? I assume we can just keep one.

File: src/event-stream/event-stream.interfaces.ts [hunk 11]
```diff
@@ -56,8 +56,13 @@ export interface Event {
   logIndex: string;
   timestamp: string;
   data: any;
-  inputMethod?: string;
-  inputArgs?: Record<string, any>;
+  inputMethod: string;
+  inputArgs: {
+    amount: string;
+    data: string;
+    to?: string;
```
Comment (awrichar) [inline src/event-stream/event-stream.interfaces.ts:63]: Maybe a nitpick, but I think the interface at this level should be generic to any event stream. If you expect particular named args I would cast that in the tokens service.

File: src/event-stream/event-stream.service.ts [hunk 12]
```diff
@@ -257,9 +264,12 @@ export class EventStreamService {
 
   async getOrCreateSubscription(
     instancePath: string,
+    methodABI: IAbiMethod,
```
Comment (awrichar) [inline src/event-stream/event-stream.service.ts:267]: I would name this `eventABI` for clarity

File: src/contractStandards/ERC20WithDataABI.ts [hunk 13]
```diff
@@ -0,0 +1,352 @@
+export const ERC20WithDataABI = [
```
Comment (awrichar) [inline src/contractStandards/ERC20WithDataABI.ts:1]: Did you convert JSON to TypeScript? Might consider just storing it as `.json` and importing the JSON file as an object, if that requires less steps. Up to you since I don't know the exact workflow to go from Solidity to here.

Comment (eberger727) [reply src/contractStandards/ERC20WithDataABI.ts:1]: Good idea, will just do `import ERC20WithDataABI from 'path/to/solidity/build/contracts/ERC20WithData.json';`

File: src/health/health.controller.ts (hunk 10 above)
Comment (eberger727) [reply src/health/health.controller.ts:33]: Not sure. @calbritt was there a reason for two ping checks?

Comment (eberger727) [reply src/health/health.controller.ts:33]: Oh, it was previously for two urls...I'll remove since we now only need one url, `baseUrl`

File: src/tokens/tokens.controller.ts [hunk 14]
```diff
@@ -36,16 +36,14 @@ export class TokensController {
...
   }
 
   @Post('createpool')
-  @HttpCode(202)
+  @HttpCode(200)
   @ApiOperation({
-    summary: 'Create a new ERC20 token contract',
-    description:
-      'Will be followed by a websocket notification with event=token-pool and data=TokenPoolEvent',
+    summary: 'Confirm and activate token pool',
```
[the remark this answers is not in the document]
Comment (eberger727) [reply src/tokens/tokens.controller.ts:46]: How about:
`summary: 'Register a new ERC20 or ERC721 contract'`
`description: 'The contract must be deployed on-chain before calling this method'`

File: src/tokens/tokens.controller.ts (hunk 5 above)
Comment (eberger727) [reply src/tokens/tokens.controller.ts:58]: Correct

File: solidity/contracts/TestERC20.sol [hunk 15]
```diff
@@ -0,0 +1,9 @@
+//SPDX-License-Identifier: Unlicense
+pragma solidity ^0.8.0;
+
+// Only used to generate ERC1155MixedFungible types for use in testing
```
Comment (awrichar) [inline solidity/contracts/TestERC20.sol:4]: Comment is wrong

File: src/event-stream/event-stream.service.ts [hunk 16]
```diff
@@ -233,18 +234,24 @@ export class EventStreamService {
 
   async createSubscription(
     instancePath: string,
+    methodABI: IAbiMethod,
```
Comment (awrichar) [inline src/event-stream/event-stream.service.ts:237]: Would suggest `eventABI` as the name here too

File: src/tokens/tokens.interfaces.ts [hunk 17]
```diff
@@ -15,14 +15,23 @@
...
-import { IsNotEmpty, IsOptional } from 'class-validator';
+import { IsDefined, IsEnum, IsNotEmpty, IsOptional } from 'class-validator';
 
 // Ethconnect interfaces
 export interface EthConnectAsyncResponse {
   sent: boolean;
   id: string;
 }
 
+export interface EthConnectContractsResponse {
```
Comment (awrichar) [inline src/tokens/tokens.interfaces.ts:26]: I think you deleted the only usage of this interface, so it can probably be removed

File: src/tokens/tokens.service.ts [hunk 18]
```diff
@@ -129,101 +155,144 @@ export class TokensService {
...
   async balance(dto: TokenBalanceQuery): Promise<TokenBalance> {
+    const validPoolId: ITokenPool = this.validatePoolId(new URLSearchParams(dto.poolId));
+    const methodAbi = this.getMethodAbi(new URLSearchParams(dto.poolId), 'BALANCE');
     const response = await lastValueFrom(
-      this.http.get<EthConnectReturn>(`${this.contractInstanceUrl}/${dto.poolId}/balanceOf`, {
-        params: {
-          account: dto.account,
+      this.http.post<EthConnectAsyncResponse>(`${this.baseUrl}`, {
+        headers: {
+          type: sendTransactionHeader,
```
Comment (awrichar) [inline src/tokens/tokens.service.ts:287]: Balance is intended to be a synchronous "call" to eth which returns the actual balance - not a submitted transaction. Feel free to completely omit it for now if you want - it's not a required part of the fftokens interface.

File: src/tokens/tokens.service.ts (hunk 6 above)
Comment (eberger727) [reply src/tokens/tokens.service.ts]: Removed

**Review by awrichar — COMMENTED** [review]
Looking really good - just a few more changes requested and then it's ready.

File: solidity/contracts/TestERC20.sol (hunk 15 above)
Comment (eberger727) [reply solidity/contracts/TestERC20.sol:4]: Removed file

File: src/tokens/tokens.service.ts (hunk 18 above)
Comment (eberger727) [reply src/tokens/tokens.service.ts]: Removed

**Comment by eberger727** [comment]
#14
